### PR TITLE
Create a ref-assembly package for HttpHandler

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/dotnet/aspnetcore/issues/19202

------------------------

This adds a package to carry a ref assembly for WebAssemblyHttpHandler. The current contents resemble what's currently in the runtime, we can update it once the HttpHandler changes are in the runtime. I choose `Microsoft.AspNetCore.Components.WebAssembly.HttpHandler` as the package name since it's meant to be a stop-gap until the runtime package is available in .NET 5.0. We can move this code over to dotnet/Blazor once we're ok with the shape of this change.

### CODE REVIEW

File: src/Components/WebAssembly/WebAssembly/src/Http/BrowserRequestCache.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        NoStore,
+
+        /// <summary>
+        /// The browser fetches the resource from the remote server without first looking in the cache,
+        /// but then will update the cache with the downloaded resource.
+        /// </summary>
+        Reload,
+
+        /// <summary>
+        /// The browser looks for a matching request in its HTTP cache.
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Http/BrowserRequestCache.cs:29]: Looks like a wrongly copy/pasted description.

File: src/Components/WebAssembly/WebAssembly/src/Microsoft.AspNetCore.Components.WebAssembly.nuspec [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Microsoft.AspNetCore.Components.WebAssembly.nuspec:1]: I should delete this. We rely on the project file to generate this.

File: src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        /// See https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream.
+        /// </remarks>
+        public static HttpRequestMessage EnableStreamingResponse(this HttpRequestMessage requestMessage)
+        {
+            if (requestMessage is null)
+            {
+                throw new ArgumentNullException(nameof(requestMessage));
+            }
+
+            requestMessage.Properties["WebAssemblyReadResponseAsStream"] = true;
```
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:164]: This string needs to be updated: `WebAssemblyEnableStreamingResponse`

File: src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        /// <summary>
+        /// Configures a value for the 'integrity' option for the HTTP request.
+        /// </summary>
+        /// <param name="requestMessage">The <see cref="HttpRequestMessage"/>.</param>
+        /// <param name="integrity">The subresource integrity descriptor.</param>
+        /// <returns>The <see cref="HttpRequestMessage"/>.</returns>
+        /// <remarks>
+        /// See https://developer.mozilla.org/en-US/docs/Web/API/Request/integrity
+        /// </remarks>
+        public static HttpRequestMessage SetBrowserIntegrity(this HttpRequestMessage requestMessage, string integrity)
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:109]:
```suggestion
        public static HttpRequestMessage SetBrowserRequestIntegrity(this HttpRequestMessage requestMessage, string integrity)
```

All the preceding methods have "Request" in the name

Comment (SteveSandersonMS) [reply src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:109]: Also helps to clarify it's not the integrity of the browser we're checking, but of this individual request

File: src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        /// Configures a value for the HTTP request.
+        /// </summary>
+        /// <param name="requestMessage">The <see cref="HttpRequestMessage"/>.</param>
+        /// <param name="name">The name of the HTTP fetch option.</param>
+        /// <param name="value">The value, which must be JSON-serializable.</param>
+        /// <returns>The <see cref="HttpRequestMessage"/>.</returns>
+        /// <remarks>
+        /// See https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
+        /// </remarks>
+        public static HttpRequestMessage SetBrowserFetchOption(this HttpRequestMessage requestMessage, string name, object value)
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:122]:
```suggestion
        public static HttpRequestMessage SetBrowserRequestOption(this HttpRequestMessage requestMessage, string name, object value)
```

Similar reason to above.

Also, in JavaScript the interface defining the options is called `RequestInit`, so it sounds right to use the term "request" rather than "fetch" from a JS perspective too.

File: src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        /// See https://developer.mozilla.org/en-US/docs/Web/API/Request/integrity
+        /// </remarks>
+        public static HttpRequestMessage SetBrowserIntegrity(this HttpRequestMessage requestMessage, string integrity)
+            => SetBrowserFetchOption(requestMessage, "integrity", integrity);
+
+        /// <summary>
+        /// Configures a value for the HTTP request.
+        /// </summary>
+        /// <param name="requestMessage">The <see cref="HttpRequestMessage"/>.</param>
+        /// <param name="name">The name of the HTTP fetch option.</param>
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:116]:
```suggestion
        /// <param name="name">The name of the option, which should correspond to a key defined on https://fetch.spec.whatwg.org/#requestinit</param>
```

Not sure if URLs in XML docs can be made clickable through any special syntax.

File: src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        /// Configures streaming response for the HTTP request.
+        /// </summary>
+        /// <param name="requestMessage">The <see cref="HttpRequestMessage"/>.</param>
+        /// <param name="streamingEnabled"><see langword="true"> if streaming is enabled; otherwise false.</param>
+        /// <returns>The <see cref="HttpRequestMessage"/>.</returns>
+        /// <remarks>
+        /// This API is only effective when the browser HTTP Fetch supports streaming.
+        /// See https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream.
+        /// </remarks>
+        public static HttpRequestMessage EnableStreamingResponse(this HttpRequestMessage requestMessage)
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:157]:
```suggestion
        public static HttpRequestMessage SetBrowserResponseStreamingEnabled(this HttpRequestMessage requestMessage, bool enabled)
```

Name suggestion is so it appears alphabetically with the others in intellisense.

Suggestion to make it possible to set "false" is because maybe in the future it might default to true, and you need a way to turn it off. Also with all the other methods here, it's possible to undo them by overwriting with `null` or similar, so it's nice if this one is undoable too.

What do you think?

File: src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs (hunk 3 above)
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Http/WebAssemblyHttpRequestMessageExtensions.cs:164]:
```suggestion
            requestMessage.Properties["WebAssemblyReadResponseAsStream"] = enabled;
```

File: src/Components/WebAssembly/testassets/StandaloneApp/Pages/FetchData.razor [hunk 8]
```diff
@@ -52,8 +52,13 @@ else
     protected override async Task OnParametersSetAsync()
     {
         startDate = StartDate.GetValueOrDefault(DateTime.Now);
-        forecasts = await Http.GetFromJsonAsync<WeatherForecast[]>(
-            $"sample-data/weather.json?date={startDate.ToString("yyyy-MM-dd")}");
+        var requestMessage = new HttpRequestMessage(HttpMethod.Get, $"sample-data/weather.json?date={startDate.ToString("yyyy-MM-dd")}");
+        requestMessage.SetBrowserRequestCache(BrowserRequestCache.NoCache);
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/testassets/StandaloneApp/Pages/FetchData.razor:56]: Please don't use `Components/WebAssembly/testassets/StandaloneApp` for testing or demonstrating random features :) It's meant to stay close to the default project template, as it's really a sandbox project not a test asset.

If you're adding this for an E2E test, could you do it inside `Components\test\testassets\BasicTestApp` instead?

Comment (SteveSandersonMS) [reply src/Components/WebAssembly/testassets/StandaloneApp/Pages/FetchData.razor:56]: Also, there are a couple of redundant blank lines below.

File: src/Components/WebAssembly/testassets/StandaloneApp/_Imports.razor [hunk 9]
```diff
@@ -2,5 +2,6 @@
 @using System.Net.Http.Json
 @using Microsoft.AspNetCore.Components.Routing
 @using Microsoft.AspNetCore.Components.Web
+@using Microsoft.AspNetCore.Components.WebAssembly.Http
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/testassets/StandaloneApp/_Imports.razor:5]: If this is also going into the project template then good, or if not then please revert.

File: src/Components/test/testassets/BasicTestApp/Program.cs [hunk 10]
```diff
@@ -28,16 +27,18 @@ public static async Task Main(string[] args)
 
             var builder = WebAssemblyHostBuilder.CreateDefault(args);
 
+            var httpClient = new HttpClient();
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Create("WEBASSEMBLY")))
             {
                 // Needed because the test server runs on a different port than the client app,
                 // and we want to test sending/receiving cookies under this config
-                WebAssemblyHttpMessageHandlerOptions.DefaultCredentials = FetchCredentialsOption.Include;
+                httpClient = new HttpClient(new ConfigureForCorsHandler(new WebAssemblyHttpHandler()));
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/BasicTestApp/Program.cs:35]: Now we have the capability of doing this on a per request basis, I would have a slight preference for moving the CORS-specific test logic into the CORS-specific test and not having it apply globally to the whole project.

Do you think that's feasible?

**Comment by SteveSandersonMS** [comment]
BTW if the suggested API changes aren't possible because the refassembly has to correspond to an existing underlying implementation, totally fine to leave that until a future follow-up.

If so, and if you do agree that API changes are beneficial, would you be able to add an issue to track this?

File: src/Components/WebAssembly/WebAssembly/src/Http/BrowserRequestCache.cs (hunk 1 above)
Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Http/BrowserRequestCache.cs:30]: I've copied the description verbatim from https://developer.mozilla.org/en-US/docs/Web/API/Request/cache

File: src/Components/test/testassets/BasicTestApp/HttpClientTest/CookieCounterComponent.razor [hunk 11]
```diff
@@ -1,3 +1,5 @@
+@using Microsoft.AspNetCore.Components.WebAssembly.Http
```
Comment (pranavkm) [inline src/Components/test/testassets/BasicTestApp/HttpClientTest/CookieCounterComponent.razor:1]: Do you think this import needs to be in `_Imports.razor`? Another alternative might be to move the extensions to `Microsoft.AspNetCore.Components.WebAssembly` instead, so it's more discoverable.

File: eng/ProjectReferences.props [hunk 12]
```diff
@@ -12,6 +12,7 @@
     <ProjectReferenceProvider Include="Microsoft.AspNetCore.Components.DataAnnotations.Validation" ProjectPath="$(RepoRoot)src\Components\WebAssembly\Validation\src\Microsoft.AspNetCore.Components.DataAnnotations.Validation.csproj"  />
     <ProjectReferenceProvider Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" ProjectPath="$(RepoRoot)src\Components\WebAssembly\WebAssembly.Authentication\src\Microsoft.AspNetCore.Components.WebAssembly.Authentication.csproj"  />
     <ProjectReferenceProvider Include="Wasm.Performance.Driver" ProjectPath="$(RepoRoot)src\Components\benchmarkapps\Wasm.Performance\Driver\Wasm.Performance.Driver.csproj" RefProjectPath="$(RepoRoot)src\Components\benchmarkapps\Wasm.Performance\ref\Wasm.Performance.Driver.csproj" />
+    <ProjectReferenceProvider Include="Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler" ProjectPath="$(RepoRoot)src\Components\WebAssembly\WebAssemblyHttpHandler\src\Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler.csproj" RefProjectPath="$(RepoRoot)src\Components\WebAssembly\WebAssemblyHttpHandler\ref\Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler.csproj" />
```
Comment (javiercn) [inline eng/ProjectReferences.props:15]: typo

**Review by javiercn — COMMENTED** [review]
Looks great!

I don't have a lot of context on this area, so I'll let Steve have the final say!

**Comment by dougbu** [comment]
Nothing should have a ref/ assembly unless it's in the shared framework. We removed extra ref/ projects entirely in at least 'release/3.1' though that cleanup may not be complete in 'master' or 'blazor-wasm'. Why is a ref/ assembly necessary / wanted here?

Note ref/ assemblies only add baggage to shipped packages and a slight complication to assembly resolution on user's machines. The only reason would be hiding `public` API from user's compilers. But, we shouldn't need such complications.

Bottom line, add the src/ project but not the ref/ one.

File: src/Components/test/testassets/BasicTestApp/HttpClientTest/CookieCounterComponent.razor (hunk 11 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/BasicTestApp/HttpClientTest/CookieCounterComponent.razor:1]: Yeah, I'd definitely be fine with it going into `_Imports.razor` in the template. Then it doesn't have to occupy an especially prevalent namespace.

File: src/Components/WebAssembly/WebAssemblyHttpHandler/src/Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler.csproj [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssemblyHttpHandler/src/Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler.csproj:1]: Typo in the filename

Comment (SteveSandersonMS) [reply src/Components/WebAssembly/WebAssemblyHttpHandler/src/Microsoft.AspNetCore.Comopnents.WebAssembly.HttpHandler.csproj:1]: And various related filenames below

**Review by SteveSandersonMS — APPROVED** [review]
Excellent!

Would be great to fix the filename typos if possible

**Comment by pranavkm** [comment]
Thanks @dougbu. I realized that the project isn't a reference provider due to how it builds. Updated it and undid the changes to `ProjectReferences.props`

**Comment by dougbu** [comment]
@pranavkm for future reference, there's a difference between being a reference provider (which means `@(Reference>` items in other projects work) and providing a ref/ **assembly** (which is about shipping something to customers in our targeting pack). And, you don't need long paths like `..\..\WebAssemblyHttpHandler\src\Microsoft.AspNetCore.Components.WebAssembly.HttpHandler.csproj` in multiple places when using `@(Reference)` items.
